<font color="red">
**Instructor notes (remove for student version)**
This notebook has all the answers and should run with no errors. Each answer line will have a commented-out version next to it, which will be what the students see in their version.
</font>

In [3]:
# Standard boilerplate code; do not change.
import numpy as np
from datascience import *
from scipy.stats import chisquare, chi2_contingency

<font color="brown" size="5">1. Chi-square tests – data from Chi-sq. homework</font>

A new school district superintendent was preparing to reallocate resources for physically impaired students and wanted to know if the schools in the district differed in the distribution of physically impaired students. 
    
Do these data suggest that the distribution of physically impaired students is different at different schools? (Use the .05 significance level.) 

**Question 1.1:** Read in the data by filling in the blank. The name of the file is `"PI_students.csv"` Hint: make sure to include the double quotation marks!

In [4]:
PI_students = Table().read_table("PI_students.csv")
# PI_students = Table().read_table(...)

PI_students.show()

<font color="brown" size="4">🔢 Accessing Data: What is an Index?</font>

An **index** is the number used to data from an ordered structure (like a column of a table, or an array). You'll see indexes used everywhere in Python.

The most important rule in Python is that **indexing always starts at 0** (the first element).

| Structure | Syntax | Example |
| :--- | :--- | :--- |
| **List/Array** | `list_name[position]` | `my_list[0]` retrieves the **first** item. |

**Question 1.2:** Fill in the following code: what's the right index  to access the 2nd column of the table?

In [5]:
observed = PI_students.column(1)
# observed = PI_students.column(...)

print("observed column as an array:", observed)
print(type(observed))

observed column as an array: [4 1 6 3 7]
<class 'numpy.ndarray'>


**Question 1.3:** Get the total number of students by using `sum()` on `observed`.

In [6]:
PI_students_N = sum(observed)
# PI_students_N = sum(...)

print("Total number of students:", PI_students_N)

Total number of students: 21


**Question 1.4:** Calculate the degrees of freedom (Use this formula: Ncategories – 1)
*Note: This is a Chi-square test for goodness of fit since there is just 1 nominal variable.

In [7]:
PI_students_Ncategories = 5
df = PI_students_Ncategories - 1
# PI_students_Ncategories = ...
# df = PI_students_Ncategories - ...

print("Degrees of freedom:", df)

Degrees of freedom: 4


**Question 1.5:** Calculate the expected frequency (Hint: Use the formula: N / Ncategories).


In [8]:
expected_freq = PI_students_N / PI_students_Ncategories
# expected_freq = ... / ...

print(f"Expected Frequencies: {expected_freq}")

Expected Frequencies: 4.2


<font color="brown" size="4">✏️ Manually calculate chi square</font>

We _could_ just use Python to calculate chi square for us, but it's better for your understanding if you calculate it manually, for now. These steps are very similar to what you'd do in Excel.

**Question 1.6:** Get the array of observed frequencies. It's the second column in the table. What index would you use to access it?

In [9]:
observed_freq = PI_students.column(1)
# observed_freq = PI_students.column(...)

print(f"Observed Frequencies: {observed_freq}")

Observed Frequencies: [4 1 6 3 7]


**Question 1.7:** Define a new array that is the difference between observed and expected. You'll be using `observed_freq` (which you just defined), and the previous `expected_freq`.

Hint: when you subtract a singular number from an array, it returns a new array that applies that subtraction to _all_ the numbers. For example: [1, 2, 3] - 1 = [0, 1, 2]

In [10]:
diff_arr = observed_freq - expected_freq
# diff_arr = ... - ...

print("diff_arr:", diff_arr)

diff_arr: [-0.2 -3.2  1.8 -1.2  2.8]


**Question 1.8:** Define a new array by squaring `diff_arr`.

In [11]:
diff_arr_squared = diff_arr**2
# diff_arr_squared = ...**2

print("diff_arr_squared:", diff_arr_squared)

diff_arr_squared: [  0.04  10.24   3.24   1.44   7.84]


**Question 1.9:** Create a new array by dividing `diff_arr_squared` by `expected_freq`.

In [12]:
division_array = diff_arr_squared / expected_freq
# division_array = ... / ...

print('division_array:', division_array)

division_array: [ 0.00952381  2.43809524  0.77142857  0.34285714  1.86666667]


**Question 1.10:** Finally, sum up all the terms in `division_array`. This is your chi square value!

In [13]:
chi_square = sum(division_array)
# chi_square = sum(...)

print('Chi-square (X²):', chi_square)

Chi-square (X²): 5.42857142857


**Question 1.11:** Calculating p-value manually is outside the scope of this lab, so instead we'll use a shortcut: the `scipy.stats.chisquare` function!

Plug in two arguments: the array `observed` and the number `expected_freq`, which you both previously defined.

In [14]:
# Use these variables (printed for your convenience)
print("observed:", observed)
print("expected frequency:", expected_freq)

chi_square_statistic, p_value = chisquare(observed, expected_freq)
# chi_square_statistic, p_value = chisquare(..., ...)

print("p-value:", p_value)

observed: [4 1 6 3 7]
expected frequency: 4.2
p-value: 0.246079819708


<font color="brown" size="4">🔎🧠 Test your knowledge</font>

**Question 1.12:** Report chi square as seen in research articles. Use this format: X2(df, N) = x.xx, p = x.xxx

**Your answer here:** X2(4, 21) = 5.43, p = 0.246

<br>

**Question 1.13:** Indicate your decision to reject or fail to reject the null hypothesis if alpha is .05

**Your answer here:** Fail to reject null 

<br>

**Question 1.14:** What conclusion you make about the significance of the results?

**Your answer here:** The results are not significant

<br>

**Question 1.15:** Indicate what conclusion you make about the research hypothesis?

**Your answer here:** The study is inconclusive.

<font color="brown" size="5">2. Scenario for REM data</font>

A study was conducted on the effects of lack of REM sleep on the ability to recall a list of words. The study found that of participants deprived of normal REM sleep, 4 scored high on memory recall, 7 scored moderately, and 9 scored low. Of participants who received normal sleep, 10 scored high, 9 moderately, and 1 low. If χ2 = 9.24, do these results suggest that REM sleep is related to how people score on the memory recall? (Use the .05 significance level.)

In [15]:
# read in the data:
print('OBSERVED:')
rem_observed = Table().read_table("rem_observed.csv")
rem_observed.show()

print('EXPECTED:')
rem_expected = Table().read_table("rem_expected.csv")
rem_expected.show()

OBSERVED:


EXPECTED:


**Question 2.1:** Sum up all the numbers in the observed table. We'll call this variable `RO_n`, which stands for REM Observed, N.

Hint: which columns have actual data? It wouldn't make sense to sum up a column with just labels.

In [16]:
RO_n = sum(rem_observed.column(1)) + sum(rem_observed.column(2)) + sum(rem_observed.column(3))
# RO_n = sum(rem_observed.column(...)) + sum(rem_observed.column(...)) + sum(rem_observed.column(...))

print("Total number of participants:", RO_n)

Total number of participants: 40


**Question 2.2:** Calculate the df. Use this formula: (Ncolumn – 1) * (Nrows – 1)

In [17]:
# only count the columns/rows with actual data; labels don't count
Ncolumn = 3
Nrows = 2
# Ncolumn = ...
# Nrows = ...

df = (Ncolumn - 1) * (Nrows - 1)
# df = (... - 1) * (... - 1)

print("Degrees of freedom:", df)

Degrees of freedom: 2


**Question 2.3:** Let's calculate the p-value and chi square value. We'll do this by plugging `observed_array` into the `chi2_contingency` function. Some code is already provided.

In [18]:
# Reformat rem_observed into a format suitable for chi2_contingency
observed_array = rem_observed.select(1, 2, 3).values
print(observed_array)

chi2_statistic, p_value, dof, expected_array = chi2_contingency(observed_array)
# chi2_statistic, p_value, dof, expected_array = chi2_contingency(...)

print(f"Chi-square statistic: {chi2_statistic:.5f}")
print(f"P-value: {p_value:.5f}")

[[ 4  7  9]
 [10  9  1]]
Chi-square statistic: 9.22143
P-value: 0.00994


<font color="brown" size="4">🔎🧠 Test your knowledge</font>

**Question 2.4:** Report X2 as seen in research articles. Use this format: X2 (df, N) = x.xx, p = x.xxx

(when reporting p-value always say what it equals, but if p is less than .001 then report it as p < .001)

**Your answer here:**  X^2 (2, 40) = 9.24, p = 0.01

<br>

**Question 2.5:** Indicate your decision to reject or fail to reject the null hypothesis if alpha is .05

**Your answer here:**  reject null 

<br>

**Question 2.6:** Indicate what conclusion you make about the significance of the results?

**Your answer here:** The results are significant

<br>

**Question 2.7:** Indicate what conclusion you make about the research hypothesis?

**Your answer here:** The research hypothesis is supported. The data suggests that REM sleep is related to how people score on memory recall

<font color="brown" size="5">3. Preferred Rewards</font>

An experimental psychologist plans a study in which a crucial step is offering participants a food reward. It is important that the three food rewards be equal in appeal. Thus, a prestudy was designed in which participants were asked which of the rewards they preferred. Of the 60 participants, 16 preferred cupcakes, 26 preferred candy bars, and 18 favored dried apricots. Do these scores suggest that the different foods are differentially preferred by people in general? (Use the .01 significance level.)

In [19]:
# read in the data:
preferred_rewards = Table().read_table("preferred_rewards.csv")
preferred_rewards.show()

**Question:** Calculate N. Do this by summing up the column with the number of participants.

In [20]:
PR_observed = preferred_rewards.column(1)
# PR_observed = preferred_rewards.column(...)

# Use the variable you just defined:
PR_N = sum(PR_observed)
# PR_N = sum(...)

print("Total number of participants:", PR_N)

Total number of participants: 60


**Question:** Calculate the DF. Use this formula: Ncategories – 1 (*Note: This is a Chi-square test for goodness of fit since there is just 1 nominal variable
)

**Answer:** df = 2

**Question:** Calculate the expected frequency. Use the formula: N/Ncategories

In [21]:
PR_EF = PR_N / 3
# PR_EF = ... / ...

print('Expected frequency', PR_EF)

Expected frequency 20.0


**Question:** Plug `PR_observed` into the `chisquare()` function in order to calculate chi square.

In [22]:
PR_chi2, PR_p_value = chisquare(PR_observed)
# PR_chi2, PR_p_value = chisquare(...)

print(f"Chi-square: {PR_chi2}")
print(f"p-value: {PR_p_value}")

Chi-square: 2.8000000000000003
p-value: 0.24659696394160638


<font color="brown" size="4">🔎🧠 Test your knowledge</font>

**Question:** Report X2 as seen in research articles: **X2 (df, N) = x.xx, p = x.xxx**

Hint: when reporting p-value always say what it equals, but if p is less than .001 then report it as p < .001

**Answer:** X^2 (4, 21) = 5.43, p=0.246

<br>

**Question:** Indicate what decision you make if alpha is .01. 

**Answer:** Fail to reject null 

<br>

**Question:** Indicate what conclusion you make about the significance of the results?

**Answer:** The results are not significant

<br>

**Question:** Indicate what conclusion you make about the research hypothesis?

**Answer:** The study is inconclusive.

<font color="brown" size="5">4. Scenario: Eat out vs. Show type</font>

An advertising firm wanted to target television advertisements for people who eat out often. The firm conducted a study in which 75 randomly selected people noted what they watched for a week, and then they were categorized according to the type of show they watched most. They also completed a questionnaire about how often they eat out, and were divided into those that eat out and rarely eat out. The results were that those in the group that eat out included 3 who most watched quiz shows, 9 who most watched situation comedies, 8 who most watched movies, 7 who most watched news shows, and 3 who most watched soap operas. Of those who ate out rarely, the figures were 8 quiz shows, 5 situation comedies, 13 movies, 3 news shows, and 16 soap operas. Do these data suggest that the distribution of type of shows watched is related to how often people eat out? **(Use the .01 significance level.)**

In [27]:
# read in the data
eat_observed = Table().read_table("eat_out_and_shows_observed.csv")
print('--- OBSERVED: ---')
eat_observed.show()

eat_expected = Table().read_table("eat_out_and_shows_expected.csv")
print('\n--- EXPECTED: ---')
eat_expected.show()

--- OBSERVED: ---



--- EXPECTED: ---


**Question 4.1:** Calculate N of the observed frequency table. What are the right indexes to use?

In [37]:
eat_N = sum(eat_observed.column(1) 
           + eat_observed.column(2)
           + eat_observed.column(3)
           + eat_observed.column(4)
           + eat_observed.column(5)
)

# eat_N = sum(eat_observed.column(...) 
#            + eat_observed.column(...)
#            + eat_observed.column(3)
#            + eat_observed.column(...)
#            + eat_observed.column(...)
# )

print('N:', eat_N)

N: 75


**Question 4.2:** Calculate the df.
Use this formula: (Ncolumn – 1) * (Nrows – 1)

In [25]:
eat_df = (5 - 1) * (2 - 1)
# eat_df = (... - 1) * (... - 1)

print('df:', eat_df)

df: 4


**Question 4.3:** Similar to when you defined `eat_N`, complete the line by entering in the correct column indexes. Hint: 0 is not one of them. Why is that?

In [26]:
observed_array = eat_observed.select(1, 2, 3, 4, 5).values
# observed_array = eat_observed.select(..., ..., ..., ..., ...).values

# Does some magic. Do not change.
EOAS_chi2_statistic, EOAS_p_value, dof, eat_expected_array = chi2_contingency(observed_array)

print(f"Chi-square statistic: {EOAS_chi2_statistic}")
print(f"P-value: {EOAS_p_value:}")

Chi-square statistic: 12.604997341839447
P-value: 0.013376148808778916


<font color="brown" size="4">🔎🧠 Test your knowledge</font>

**Question 4.4:** Report X2 as seen in research articles. Use this format: X2 (df, N) = x.xx, p = x.xxx

*when reporting p-value always say what it equals, but if p is less than .001 then report it as p < .001

**Answer:** X^2 (4, 75) = 12.61, p=.013

<br>

**Question 4.5:** Indicate what decision you make if alpha is .01

**Answer:** Fail to reject null

<br>

**Question 4.6:** Indicate what conclusion you make about the significance of the results?

**Answer:** The results are not significant

<br>

**Question 4.7:** Indicate what conclusion you make about the research hypothesis?

**Answer:** The study is inconclusive.